### PR TITLE
Add basic ingress support

### PR DESCRIPTION
**Pull request by andrew-su**
This adds basic support to create an ingress object per deployer.

In this implementation it uses the default domain of `example.com` so all ingress host names will be of the form `<deployername>.example.com`. Follow-up PR for changing/setting the domain.

Relevant Issue: https://github.com/projectriff/system/issues/142

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@andrew-su Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=projectriff/system&pullRequestId=150)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=projectriff/system&pullRequestId=150) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@andrew-su Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=projectriff/system&pullRequestId=150)!

File: pkg/controllers/core/deployer_controller.go [hunk 1]
```diff
@@ -132,6 +138,14 @@ func (r *DeployerReconciler) reconcile(ctx context.Context, log logr.Logger, dep
 	deployer.Status.ServiceName = childService.Name
 	deployer.Status.PropagateServiceStatus(&childService.Status)
 
+	ingress, err := r.reconcileIngress(ctx, log, deployer, childService.Name)
+	if err != nil {
+		log.Error(err, "unable to reconcile Ingress", "deployer", deployer)
+		return ctrl.Result{}, err
+	}
+
+	deployer.Status.IngressName = ingress.Name
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: The status of the ingress should be reflected on the deployer, and contribute to the deployer's ready condition

```suggestion
	deployer.Status.IngressName = ingress.Name
	deployer.Status.PropagateIngressStatus(&childIngress.Status)
```

File: pkg/controllers/core/deployer_controller.go [hunk 2]
```diff
@@ -331,6 +345,69 @@ func (r *DeployerReconciler) constructPodSpecForDeployer(deployer *corev1alpha1.
 	return podSpec
 }
 
+func (r *DeployerReconciler) reconcileIngress(ctx context.Context, log logr.Logger, deployer *corev1alpha1.Deployer, serviceName string) (*v1beta1.Ingress, error) {
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: group all of the ingress related methods together, like the deployment related methods are grouped and the service related method are grouped.

File: pkg/controllers/core/deployer_controller.go [hunk 3]
```diff
@@ -20,22 +20,25 @@ import (
 	"context"
 	"fmt"
 
-	"github.com/go-logr/logr"
-	"github.com/google/go-cmp/cmp"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/networking/v1beta1"
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: `v1beta1` isn't a meaningful package name

```suggestion
	networkingv1beta1 "k8s.io/api/networking/v1beta1"
```

File: pkg/controllers/core/deployer_controller.go [hunk 4]
```diff
@@ -20,22 +20,25 @@ import (
...
 	"k8s.io/apimachinery/pkg/api/equality"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/intstr"
+
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: rm

File: pkg/controllers/core/deployer_controller.go [hunk 5]
```diff
@@ -20,22 +20,25 @@ import (
...
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
+	"github.com/go-logr/logr"
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: rm leading blank line, which will cause these to be sorted back towards the top of the list.

File: pkg/controllers/core/deployer_controller.go [hunk 6]
```diff
@@ -396,11 +473,55 @@ func (r *DeployerReconciler) reconcileChildService(ctx context.Context, log logr
...
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:       labels,
+			Annotations:  make(map[string]string),
+			GenerateName: fmt.Sprintf("%s-deployer-", deployer.Name),
+			Namespace:    deployer.Namespace,
+		},
+		Spec: v1beta1.IngressSpec{
+			Rules: []v1beta1.IngressRule{
+				{
+					Host: fmt.Sprintf("%s.%s", deployer.Name, domain),
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]:
```suggestion
					Host: fmt.Sprintf("%s.%s.%s", deployer.Name, deployer.Namespace, domain),
```

File: pkg/controllers/core/deployer_controller.go [hunk 7]
```diff
@@ -396,11 +473,55 @@ func (r *DeployerReconciler) reconcileChildService(ctx context.Context, log logr
...
+		equality.Semantic.DeepEqual(desiredIngress.ObjectMeta.Labels, ingress.ObjectMeta.Labels)
+}
+
 func (r *DeployerReconciler) serviceSemanticEquals(desiredService, service *corev1.Service) bool {
 	return equality.Semantic.DeepEqual(desiredService.Spec, service.Spec) &&
 		equality.Semantic.DeepEqual(desiredService.ObjectMeta.Labels, service.ObjectMeta.Labels)
 }
 
+func (r *DeployerReconciler) constructIngressForDeployer(deployer *corev1alpha1.Deployer) (*v1beta1.Ingress, error) {
+	labels := r.constructLabelsForDeployer(deployer)
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: Only create the ingress if the service is created. We'll also want to make ingress opt-in at some point.

```suggestion
	if deployer.Status.ServiceName == "" {
		return nil, nil
	}

	labels := r.constructLabelsForDeployer(deployer)
```

File: pkg/controllers/core/deployer_controller.go [hunk 8]
```diff
@@ -396,11 +473,55 @@ func (r *DeployerReconciler) reconcileChildService(ctx context.Context, log logr
...
+			Annotations:  make(map[string]string),
+			GenerateName: fmt.Sprintf("%s-deployer-", deployer.Name),
+			Namespace:    deployer.Namespace,
+		},
+		Spec: v1beta1.IngressSpec{
+			Rules: []v1beta1.IngressRule{
+				{
+					Host: fmt.Sprintf("%s.%s", deployer.Name, domain),
+					IngressRuleValue: v1beta1.IngressRuleValue{
+						HTTP: &v1beta1.HTTPIngressRuleValue{
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: for ingress controllers that support https, do we need to define a rule, or will they handle that automatically?

It's also fine to punt on this.

**Comment by shashwathi** [comment]
@scothis : I have addressed most of your comments. Please take another look at this PR and let me know if I missed anything

**Comment by shashwathi** [comment]
@scothis : 
Updated deployer to have `ingressName`, `ingressStatus` and it does not depend on deployment Read status. Tested it locally with nginx controller

File: pkg/apis/core/v1alpha1/deployer_types.go [hunk 9]
```diff
@@ -66,12 +66,12 @@ type DeployerStatus struct {
 	// Important: Run "make" to regenerate code after modifying this file
 
 	apis.Status `json:",inline"`
-
 	// LatestImage is the most recent image resolved from the build
```
Comment (scothis) [inline pkg/apis/core/v1alpha1/deployer_types.go]: nit: restore the newline

File: pkg/controllers/core/deployer_controller.go [hunk 10]
```diff
@@ -132,8 +137,18 @@ func (r *DeployerReconciler) reconcile(ctx context.Context, log logr.Logger, dep
...
 	deployer.Status.PropagateServiceStatus(&childService.Status)
 
-	deployer.Status.ObservedGeneration = deployer.Generation
+	// reconcile ingress
+	childIngress, err := r.reconcileIngress(ctx, log, deployer, childService.Name)
+	if err != nil {
+		log.Error(err, "unable to reconcile Ingress", "deployer", deployer)
+		return ctrl.Result{}, err
+	}
+	if childIngress != nil {
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: Also need to update the status is the ingress is nil. If for some reason the ingress previously existed and now no longer needs to exists, that change should be reflected on the Deployer's status.

File: pkg/controllers/core/deployer_controller.go [hunk 11]
```diff
@@ -132,8 +137,18 @@ func (r *DeployerReconciler) reconcile(ctx context.Context, log logr.Logger, dep
...
+		deployer.Status.PropagateIngressStatus(childIngress.Status)
+		deployer.Status.IngressName = childIngress.GetName()
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: Setting the name before calling PropagateStatus is more common. Sometimes the Propagate method needs access to the resources name.

```suggestion
		deployer.Status.IngressName = childIngress.Name
```

File: pkg/controllers/core/deployer_controller.go [hunk 12]
```diff
@@ -331,6 +346,112 @@ func (r *DeployerReconciler) constructPodSpecForDeployer(deployer *corev1alpha1.
...
+		// this shouldn't happen, delete everything to a clean slate
+		for _, extraIngress := range childIngresses.Items {
+			log.Info("deleting extra ingress", "ingress", extraIngress)
+			if err := r.Delete(ctx, &extraIngress); err != nil {
+				return nil, err
+			}
+		}
+	}
+	// construct ingress if service is present
+	if deployer.Status.ServiceName == "" {
```
Comment (scothis) [inline pkg/controllers/core/deployer_controller.go]: should we move this check inside the constructIngressForDeployer method? That way we can continue the reconciliation to remove the ingress if the service name gets unset.

Comment (shashwathi) [reply pkg/controllers/core/deployer_controller.go]: Updated the status to `False` with helpful msg & reason.

File: pkg/controllers/core/deployer_controller.go (hunk 10 above)
Comment (shashwathi) [reply pkg/controllers/core/deployer_controller.go]: Updated the status to False with helpful msg & reason

**Comment by shashwathi** [comment]
@scothis : Ready for another round of review. Thanks 👍

File: pkg/controllers/core/deployer_controller.go (hunk 10 above)
Comment (scothis) [reply pkg/controllers/core/deployer_controller.go]: Take a look at https://github.com/projectriff/system/blob/8c95a169d0f915e5d2f2796127ddb28a214f9cd0/pkg/controllers/build/application_controller.go#L114-L123

**Comment by shashwathi** [comment]
@scothis  I have updated the logic to clear ingress name if there is no need for ingress. Ready for another review 👍 
btw will you merge this PR or should I merge this PR myself after you approve the above changes?

**Pull request merged**